sdk/storage/Azure.Storage.Blobs/src/BlobClient.cs

```diff
@@ -387,14 +392,136 @@ internal BlobClient(Uri blobUri, HttpPipeline pipeline)
             IProgress<StorageProgress> progressHandler = default,
             CancellationToken cancellationToken = default) =>
-            await new BlockBlobClient(this.Uri, this.Pipeline)
-                .UploadAsync(
+            await this.StagedUploadAsync(
+                content,
+                blobHttpHeaders,
+                metadata,
+                blobAccessConditions,
+                progressHandler,
+                async: true,
+                cancellationToken: cancellationToken)
+                .ConfigureAwait(false);
+
+        /// <summary>
+        /// The <see cref="StagedUploadAsync"/> operation will create a new
+        /// block blob of arbitrary size by uploading it as indiviually staged
+        /// blocks if it's larger than the
+        /// <paramref name="singleBlockThreshold"/>.
+        /// </summary>
+        /// <param name="content">
+        /// A <see cref="Stream"/> containing the content to upload.
+        /// </param>
+        /// <param name="blobHttpHeaders">
+        /// Optional standard HTTP header properties that can be set for the
+        /// block blob.
+        /// </param>
+        /// <param name="metadata">
+        /// Optional custom metadata to set for this block blob.
+        /// </param>
+        /// <param name="blobAccessConditions">
+        /// Optional <see cref="BlobAccessConditions"/> to add conditions on
+        /// the creation of this new block blob.
+        /// </param>
+        /// <param name="progressHandler">
+        /// Optional <see cref="IProgress{StorageProgress}"/> to provide
+        /// progress updates about data transfers.
+        /// </param>
+        /// <param name="singleBlockThreshold">
+        /// The maximum size stream that we'll upload as a single block.  The
+        /// default value is 256MB.
+        /// </param>
+        /// <param name="blockSize">
+        /// The size of individually staged blocks.  The default value is 4MB.
+        /// </param>
+        /// <param name="async">
+        /// </param>
+        /// <param name="cancellationToken">
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Task{Response{BlobContentInfo}}"/> describing the
+        /// state of the updated block blob.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="StorageRequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        internal async Task<Response<BlobContentInfo>> StagedUploadAsync(
+            Stream content,
+            BlobHttpHeaders? blobHttpHeaders,
+            Metadata metadata,
+            BlobAccessConditions? blobAccessConditions,
+            IProgress<StorageProgress> progressHandler,
+            long singleBlockThreshold = BlockBlobClient.BlockBlobMaxUploadBlobBytes,
```

Comment:
Do we cap singleBlockThreshold, or permit a value larger than BlockBlobMaxUploadBlobBytes?

Reply:
It's an internal API only used for testing.  I added an assertion to make sure we never go above `BlockBlobMaxUploadBlobBytes` in `BlobClient`.